File: go/src/2020/aug/word-dictionary.go
package aug

type WordDictionary struct {
	m map[byte]*WordDictionary
}

func NewWordDictionary() WordDictionary {
	return WordDictionary{
		m: make(map[byte]*WordDictionary),
	}
}

func (wd *WordDictionary) AddWord(word string) {
	current := wd
	for i := 0; i < len(word); i++ {
		w := word[i]
		if v, ok := current.m[w]; ok {
			current = v
		} else {
			current.m[w] = &WordDictionary{m: make(map[byte]*WordDictionary)}
			current = current.m[w]
		}
	}
	// represents the end of the string
	current.m['$'] = &WordDictionary{m: make(map[byte]*WordDictionary)}
}

func (wd *WordDictionary) Search(word string) bool {
	return search(word, wd)
}

func search(word string, current *WordDictionary) bool {
	if len(word) == 0 {
		return false
	}
	if word[0] == '.' {
		for _, v := range current.m {
			if v.m['$'] != nil && len(word) == 1 {
				return true
			}
			if search(word[1:], v) {
				return true
			}
		}
	} else {
		if v, ok := current.m[word[0]]; ok {
			if v.m['$'] != nil && len(word) == 1 {
				return true
			} else {
				return search(word[1:], v)
			}
		}
	}
	return false
}

File: python/2020/aug/word-dictionary.py
class WordDictionary:

    def __init__(self):
        self.root = {}

    def addWord(self, word):
        node = self.root
        for char in word:
            node = node.setdefault(char, {})
        node['$'] = None

    def search(self, word):
        nodes = [self.root]
        for char in word + '$':
            nodes = [kid for node in nodes for kid in
                     ([node[char]] if char in node else
                      filter(None, node.values()) if char == '.' else [])]
        return bool(nodes)


class WordDictionary_(object):
    def __init__(self):
        self.trie = dict()

    def addWord(self, word):
        cur = self.trie
        for c in word + "$": cur = cur.setdefault(c, {})
        print(self.trie)

    def search(self, word, cur=None):
        if not cur: cur = self.trie
        return "$" in cur if not word \
            else self.search(word[1:], cur[word[0]]) if word[0] in cur \
            else False if 'a' <= word[0] <= 'z' \
            else any(self.search(word[1:], cur[c]) for c in cur if c != "$")


if __name__ == '__main__':
    wd = WordDictionary_()
    wd.addWord("hello")
    wd.addWord("halo")
